### PR TITLE
codegen: use new {re,de,}allocator annotations in llvm

### PR DESCRIPTION
**Pull request by durin42**
This obviates the patch that teaches LLVM internals about
_rust_{re,de}alloc functions by putting annotations directly in the IR
for the optimizer.

The sole test change is required to anchor FileCheck to the body of the
`box_uninitialized` method, so it doesn't see the `allocalign` on
`__rust_alloc` and get mad about the string `alloca` showing up. Since I
was there anyway, I added some checks on the attributes to prove the
right attributes got set.

r? @nikic

### CODE REVIEW

File: compiler/rustc_llvm/llvm-wrapper/RustWrapper.cpp [hunk 1]
```diff
@@ -303,6 +311,22 @@ extern "C" LLVMAttributeRef LLVMRustCreateUWTableAttr(LLVMContextRef C, bool Asy
 #endif
 }
 
+extern "C" LLVMAttributeRef LLVMRustCreateAttrStringValue(LLVMContextRef C, const char* Name, const char* Value) {
+  return wrap(Attribute::get(*unwrap(C), Name, Value));
+}
```
Comment (nikic) [inline compiler/rustc_llvm/llvm-wrapper/RustWrapper.cpp:316]: This function is unused.

File: compiler/rustc_codegen_llvm/src/attributes.rs [hunk 2]
```diff
@@ -307,10 +307,42 @@ pub fn from_fn_attrs<'ll, 'tcx>(
         to_add.push(llvm::CreateAttrStringValue(cx.llcx, "branch-target-enforcement", "false"));
     }
     if codegen_fn_attrs.flags.contains(CodegenFnAttrFlags::ALLOCATOR) {
+        if llvm_util::get_version() >= (15, 0, 0) {
+            to_add.push(llvm::CreateAttrStringValue(cx.llcx, "alloc-family", "__rust_alloc"));
+            // apply to argument place instead of function
+            let alloc_align = AttributeKind::AllocAlign.create_attr(cx.llcx);
+            attributes::apply_to_llfn(llfn, AttributePlace::Argument(1), &[alloc_align]);
+        }
+        to_add.push(llvm::CreateAllocSizeAttr(cx.llcx, 0));
```
Comment (nikic) [inline compiler/rustc_codegen_llvm/src/attributes.rs:316]: Might as well put this into the LLVM 15 branch as well? Don't think there's much point to emitting it before that, even if it exists.

File: compiler/rustc_codegen_llvm/src/attributes.rs [hunk 3]
```diff
@@ -307,10 +307,42 @@ pub fn from_fn_attrs<'ll, 'tcx>(
...
+            to_add.push(llvm::CreateAttrStringValue(cx.llcx, "alloc-family", "__rust_alloc"));
+            to_add.push(llvm::CreateAllocKindAttr(cx.llcx, 1 << 1 | 1 << 5));
+            // applies to argument place instead of function place
+            let allocated_pointer = AttributeKind::AllocatedPointer.create_attr(cx.llcx);
+            attributes::apply_to_llfn(llfn, AttributePlace::Argument(0), &[allocated_pointer]);
+            // apply to argument place instead of function
+            let alloc_align = AttributeKind::AllocAlign.create_attr(cx.llcx);
+            attributes::apply_to_llfn(llfn, AttributePlace::Argument(2), &[alloc_align]);
+        }
+        to_add.push(llvm::CreateAllocSizeAttr(cx.llcx, 3));
```
Comment (nikic) [inline compiler/rustc_codegen_llvm/src/attributes.rs:335]: Missing noalias return for reallocator?

File: compiler/rustc_llvm/llvm-wrapper/RustWrapper.cpp [hunk 4]
```diff
@@ -303,6 +311,22 @@ extern "C" LLVMAttributeRef LLVMRustCreateUWTableAttr(LLVMContextRef C, bool Asy
...
+
+extern "C" LLVMAttributeRef LLVMRustCreateAllocSizeAttr(LLVMContextRef C, uint32_t ElementSizeArg) {
+  return wrap(Attribute::getWithAllocSizeArgs(*unwrap(C), ElementSizeArg, None));
+}
+
+extern "C" LLVMAttributeRef LLVMRustCreateAllocKindAttr(LLVMContextRef C, uint64_t AllocKindArg) {
+#if LLVM_VERSION_GE(15, 0)
+  return wrap(Attribute::get(*unwrap(C), Attribute::AllocKind, AllocKindArg));
+#else
+  return wrap(Attribute());
```
Comment (nikic) [inline compiler/rustc_llvm/llvm-wrapper/RustWrapper.cpp:326]: I'd prefer a `report_fatal_error()` here.

File: library/alloc/src/alloc.rs [hunk 5]
```diff
@@ -30,8 +30,10 @@ extern "Rust" {
     #[rustc_allocator]
     #[rustc_allocator_nounwind]
     fn __rust_alloc(size: usize, align: usize) -> *mut u8;
```
Comment (nikic) [inline library/alloc/src/alloc.rs:32]: We should also annotate the alloc_zeroed below (with zeroed allockind).

File: compiler/rustc_llvm/llvm-wrapper/RustWrapper.cpp (hunk 1 above)
Comment (durin42) [reply compiler/rustc_llvm/llvm-wrapper/RustWrapper.cpp]: Oops. Thanks!

File: compiler/rustc_codegen_llvm/src/attributes.rs (hunk 2 above)
Comment (durin42) [reply compiler/rustc_codegen_llvm/src/attributes.rs]: Sure.

File: compiler/rustc_codegen_llvm/src/attributes.rs (hunk 3 above)
Comment (durin42) [reply compiler/rustc_codegen_llvm/src/attributes.rs]: Done.

File: compiler/rustc_llvm/llvm-wrapper/RustWrapper.cpp (hunk 4 above)
Comment (durin42) [reply compiler/rustc_llvm/llvm-wrapper/RustWrapper.cpp]: Done.

File: library/alloc/src/alloc.rs (hunk 5 above)
Comment (durin42) [reply library/alloc/src/alloc.rs:32]: Done, but I'm uneasy per a comment in a bug about "weird UB-like behaviour with bitvec iterators in rustc_data_structures" last time we did this. It's a separate commit so it'll be easy to back it out if we see the problem, but maybe we should even do it as a separate PR to isolate possible blast radius?

Comment (nikic) [reply library/alloc/src/alloc.rs:32]: I'm not particularly concerned about this, given how much time has passed. I have been testing rust with `__rust_alloc_zeroed` registered as a TLI allocator (before your patches...) and didn't encounter any issues.

File: compiler/rustc_codegen_llvm/src/attributes.rs [hunk 6]
```diff
@@ -306,11 +306,56 @@ pub fn from_fn_attrs<'ll, 'tcx>(
...
+    if codegen_fn_attrs.flags.contains(CodegenFnAttrFlags::ALLOCATOR)
+        || codegen_fn_attrs.flags.contains(CodegenFnAttrFlags::ALLOCATOR_ZEROED)
+    {
+        if llvm_util::get_version() >= (15, 0, 0) {
+            to_add.push(llvm::CreateAttrStringValue(cx.llcx, "alloc-family", "__rust_alloc"));
+            // apply to argument place instead of function
+            let alloc_align = AttributeKind::AllocAlign.create_attr(cx.llcx);
+            attributes::apply_to_llfn(llfn, AttributePlace::Argument(1), &[alloc_align]);
+        }
+        if llvm_util::get_version() >= (15, 0, 0) {
```
Comment (nikic) [inline compiler/rustc_codegen_llvm/src/attributes.rs:318]: Can combine these two branches, the condition is the same.

File: compiler/rustc_codegen_llvm/src/attributes.rs [hunk 7]
```diff
@@ -306,11 +306,56 @@ pub fn from_fn_attrs<'ll, 'tcx>(
         // Need this for AArch64.
         to_add.push(llvm::CreateAttrStringValue(cx.llcx, "branch-target-enforcement", "false"));
     }
-    if codegen_fn_attrs.flags.contains(CodegenFnAttrFlags::ALLOCATOR) {
+    if codegen_fn_attrs.flags.contains(CodegenFnAttrFlags::ALLOCATOR)
+        || codegen_fn_attrs.flags.contains(CodegenFnAttrFlags::ALLOCATOR_ZEROED)
+    {
+        if llvm_util::get_version() >= (15, 0, 0) {
+            to_add.push(llvm::CreateAttrStringValue(cx.llcx, "alloc-family", "__rust_alloc"));
```
Comment (nikic) [inline compiler/rustc_codegen_llvm/src/attributes.rs:313]: I'd suggest extracting a small `create_alloc_family_attr()` function, so we need to hardcode the `__rust_alloc` family only once.

File: compiler/rustc_llvm/llvm-wrapper/RustWrapper.cpp [hunk 8]
```diff
@@ -303,6 +311,66 @@ extern "C" LLVMAttributeRef LLVMRustCreateUWTableAttr(LLVMContextRef C, bool Asy
...
+};
+
+inline LLVMRustAllocKindFlags operator&(LLVMRustAllocKindFlags A, LLVMRustAllocKindFlags B) {
+  return static_cast<LLVMRustAllocKindFlags>(static_cast<uint64_t>(A) &
+                                      static_cast<uint64_t>(B));
+}
+
+inline bool isSet(LLVMRustAllocKindFlags F) { return F != LLVMRustAllocKindFlags::Unknown; }
+
+#if LLVM_VERSION_GE(15, 0)
```
Comment (nikic) [inline compiler/rustc_llvm/llvm-wrapper/RustWrapper.cpp:339]: The isSet() function should probably also be behind the guard?

File: compiler/rustc_llvm/llvm-wrapper/RustWrapper.cpp [hunk 9]
```diff
@@ -303,6 +311,66 @@ extern "C" LLVMAttributeRef LLVMRustCreateUWTableAttr(LLVMContextRef C, bool Asy
...
+  Zeroed = 1 << 4,
+  Aligned = 1 << 5,
+};
+
+inline LLVMRustAllocKindFlags operator&(LLVMRustAllocKindFlags A, LLVMRustAllocKindFlags B) {
+  return static_cast<LLVMRustAllocKindFlags>(static_cast<uint64_t>(A) &
+                                      static_cast<uint64_t>(B));
+}
+
+inline bool isSet(LLVMRustAllocKindFlags F) { return F != LLVMRustAllocKindFlags::Unknown; }
```
Comment (nikic) [inline compiler/rustc_llvm/llvm-wrapper/RustWrapper.cpp:337]: inline -> static. Non-static inline has semantics I'd rather not think about.

**Review by nikic — COMMENTED** [review]
As you've found out, we should have a separate test with min-llvm-version for testing the attributes.

**Comment by durin42** [comment]
> As you've found out, we should have a separate test with min-llvm-version for testing the attributes.

I ended up making "llvm14" variants of the tests that don't have any llvm-15 specific stuff, and marking the "main" tests as requiring LLVM 15, with an eye towards deleting the old-llvm version and only really having to think about the LLVM 15 flavor fairly soon.

Also, we have zero tests that emit __rust_realloc, but I have no idea how one would make that happen.

File: compiler/rustc_codegen_llvm/src/attributes.rs [hunk 10]
```diff
@@ -224,6 +224,10 @@ pub(crate) fn default_optimisation_attrs<'ll>(
     attrs
 }
 
+fn create_alloc_family_attr<'ll>(llcx: &'ll llvm::Context) -> &'ll llvm::Attribute {
```
Comment (nikic) [inline compiler/rustc_codegen_llvm/src/attributes.rs:227]: You should be able to elide the explicit lifetimes here.

File: compiler/rustc_middle/src/middle/codegen_fn_attrs.rs [hunk 11]
```diff
@@ -91,6 +91,12 @@ bitflags! {
         const NO_COVERAGE               = 1 << 15;
         /// `#[used(linker)]`: indicates that LLVM nor the linker can eliminate this function.
         const USED_LINKER               = 1 << 16;
+        /// `#[rustc_deallocator]`: a hint to LLVM that the function only deallocates memory.
+        const DEALLOCATOR               = 1 << 17;
+        /// `#[rustc_reallocator]`: a hint to LLVM that the function only reallocates memory.
+        const REALLOCATOR               = 1 << 18;
+        /// `#[rustc_allocator_zeroed]`: a hint to LLVM that the function only allocates zeroed memory.
+        const ALLOCATOR_ZEROED = 1<<19;
```
Comment (nikic) [inline compiler/rustc_middle/src/middle/codegen_fn_attrs.rs:99]: nit: Doesn't follow the alignment of the other lines.

**Review by nikic — APPROVED** [review]
Looks good to me. Can you please squash the commits?

**Comment by durin42** [comment]
Squashed as requested

**Comment by nikic** [comment]
Also needs a rebase :)

**Comment by durin42** [comment]
ugh, I thought I had. I must have held git wrong somehow.

**Comment by durin42** [comment]
Updated to not require i64 sizes in the allocator functions. Should pass on both 32 and 64 bit now. 😄

**Pull request merged**